<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Devoir5_MappingVCF_English</title>
  <link rel="stylesheet" href="https://stackedit.io/style.css" />
</head>

<body class="stackedit">
  <div class="stackedit__html"><h1 id="devoir-5">Devoir 5</h1>
<p>Download the following files for today’s exercises</p>
<pre><code>wget https://raw.githubusercontent.com/nomascus/ANT3475_2026/refs/heads/main/FILES/Mapping_VCF/reference_mt.fasta
wget https://raw.githubusercontent.com/nomascus/ANT3475_2026/refs/heads/main/FILES/Mapping_VCF/mt_multiSNP_2000_R1.fastq 
wget https://raw.githubusercontent.com/nomascus/ANT3475_2026/refs/heads/main/FILES/Mapping_VCF/mt_multiSNP_2000_R2.fastq
</code></pre>
<h2 id="file-descriptions">File Descriptions</h2>

<table>
<thead>
<tr>
<th>file name</th>
<th>file type</th>
<th>file description</th>
</tr>
</thead>
<tbody>
<tr>
<td>reference_mt.fasta</td>
<td>FASTA</td>
<td>human mitochondrial reference genome</td>
</tr>
<tr>
<td>mt_multiSNP_2000_R1.fastq</td>
<td>FASTQ</td>
<td>2000 human mitochondrial reads (Forward reads)</td>
</tr>
<tr>
<td>mt_multiSNP_2000_R1.fastq</td>
<td>FASTQ</td>
<td>2000 human mitochondrial reads (Reverse reads)</td>
</tr>
</tbody>
</table><ol>
<li>
<p>Load the modules for bwa, samtools, and bcftools</p>
</li>
<li>
<p>Make all the index files for the reference genome</p>
</li>
<li>
<p>Map the reads to the reference genome</p>
</li>
<li>
<p>Compress, sort and index the SAM file</p>
</li>
<li>
<p>Filter the BAM File to remove reads with mapping scores less than 20</p>
</li>
<li>
<p>Call variants (make a VCF) from the filtered BAM file</p>
</li>
<li>
<p>Filter the VCF to remove SNPs with quality scores less than 30</p>
</li>
<li>
<p>Which SNP was removed from the VCF during the filtration process in step 7?</p>
</li>
</ol>
</div>
</body>

</html>
